TAO/TAO_IDL/fe/idl.ypp

```diff
@@ -6381,52 +6468,53 @@ event_decl :
         event_header
         event_rest_of_header
         {
-          UTL_Scope *s = idl_global->scopes ().top_non_null ();
-          AST_EventType *e = 0;
-          AST_Interface *i = 0;
+          UTL_Scope *scope = idl_global->scopes ().top_non_null ();
+          Identifier *&event_id = $1;
+          FE_OBVHeader *&event_header = $2;
+          AST_EventType *eventtype = 0;
 
-          if (s != 0 && $1 != 0)
+          if (scope && event_header)
             {
               // We create the scoped name here instead of with the
-              // FE_EventHeader because there is a token for it only here.
-              UTL_ScopedName sn ($1,
-                                 0);
-              e =
-                idl_global->gen ()->create_eventtype (
-                    &sn,
-                    $2->inherits (),
-                    $2->n_inherits (),
-                    $2->inherits_concrete (),
-                    $2->inherits_flat (),
-                    $2->n_inherits_flat (),
-                    $2->supports (),
-                    $2->n_supports (),
-                    $2->supports_concrete (),
-                    false,
-                    $2->truncatable (),
-                    false
-                  );
-              i = AST_Interface::narrow_from_decl (e);
-              AST_Interface::fwd_redefinition_helper (i,
-                                                      s);
+              // FE_OBVHeader because there is a token for it only here.
+              UTL_ScopedName sn (event_id, 0);
+              eventtype = idl_global->gen ()->create_eventtype (
+                &sn,
+                event_header->inherits (),
+                event_header->n_inherits (),
+                event_header->inherits_concrete (),
+                event_header->inherits_flat (),
+                event_header->n_inherits_flat (),
+                event_header->supports (),
+                event_header->n_supports (),
+                event_header->supports_concrete (),
+                false,
+                event_header->truncatable (),
+                false);
+              AST_Interface *eventtype_as_interface =
+                dynamic_cast<AST_Interface *> (eventtype);
+              AST_Interface::fwd_redefinition_helper (
+                eventtype_as_interface, scope);
+
               /*
                * Add the eventetype to its definition scope
                */
-              e = AST_EventType::narrow_from_decl (i);
-              (void) s->fe_add_eventtype (e);
+              scope->fe_add_eventtype (eventtype);
```

Comment:
I looks like I might have make the same mistake here as for `valuetype`, thinking `e` from before is the same as `i`. I don't know if this is causing the issue you're looking at @jwillemsen, but I change it to do what it was doing before.

Reply:
Just compile the unit test that is failing, that shows the problem

Reply:
The failing test isn't a unit test, but I reproduced the failure. I was confused before because I used this PR as a reference in my previous comments for the line numbers, but this was wrong because of my other PR.

I fixed this issue with `eventtype` and also with `component` and was able to run `tao_idl3_to_idl2` on the failing file. I will open a PR after my computer is done building ACE through CIAO.